### ISSUE TITLE
How to use environment variable for basicauth user ?

### ISSUE THREAD

**Issue by hollowshiroyuki** [lucaslorentz/caddy-docker-proxy#466]
Hello !

Is there a way to set the user for the basicauth directive with an environment variable ?

The usual syntax would be :
```
caddy.basicauth: "* bcrypt"
caddy.basicauth.admin: "${HASHED_PASSWORD}"
```
But the yaml keys in docker compose doesn't resolve environment so :
`caddy.basicauth.${USER}: "${HASHED_PASSWORD}"` won't work.

Isn't there a syntax where we can "hardcode" the block or part of it like :
```
caddy.basicauth: "* bcrypt { ${USER} ${HASHED_PASSWORD} }"
```
Thanks for your assistance !
Best regards, Hollow

**Comment by IndeedNotJames** [comment]
As you already found out, variable substitution/interpolation in docker-compose does only work for values, not keys.

You could, however, use snippets and `{args.*}`with [`import`](https://caddyserver.com/docs/caddyfile/directives/import) instead.
See https://caddyserver.com/docs/caddyfile/concepts#snippets

Let me know if I should provide you with an example if the docs are unclear :)

**Comment by hollowshiroyuki** [comment]
Thanks for the suggestion !
I thought about using snippets but didn't know about `{args.*}`.

But now I have a weird behavior with this :
```yml
caddy_0: "(basic_auth)"
caddy_0.basicauth: "* bcrypt"
"caddy_0.basicauth.{args.0}": "${HASHED_PASSWORD}"

caddy_1: "domain"
caddy_1.import: "basic_auth ${USER}"
caddy_1.reverse_proxy: "container:80"
```
The logs outputs this :
```json
{
    "level": "info",
    "ts": 1677263426.517001,
    "logger": "docker-proxy",
    "msg": "New Caddyfile",
    "caddyfile": "(basic_auth) {\n\tbasicauth * bcrypt {\n\t\t{args {\n\t\t\t0} [REDACTED]\n\t\t}\n\t}\n}\n"
}
```
Am I doing something wrong or is it a parser issue ?

**Comment by francislavoie** [comment]
Yeah, that certainly looks like a bug. It's adding newlines to the args placeholder. It shouldn't do that when there's no newline or space after the `{`.

As a workaround, you can use `CADDY_DOCKER_CADDYFILE_PATH` to set up a base Caddyfile which you can mount in the Caddy container which can contain snippets, then you can use labels to import that snippet.

**Comment by hollowshiroyuki** [comment]
I'm using docker in combination with ansible and I was using `caddy-docker-proxy` to avoid copying a Caddyfile for a simple configuration.
I'm just gonna use a Caddyfile templated with jinja and ansible for now. 😛

Anyways, thanks for the help !

**Comment by zieddhf** [comment]
Hello !
I'm actually trying to set environment variables for the basicauth user and password in my Caddyfile

the syntax is the following : 

```
 reverse_proxy * localhost:3000

  basicauth /* {
       {$GRAFANA_USER} {$GRAFANA_PWD}
  }
```

I tried export command to set the environment variables but the values return empty, the only way i found is by adding Environment directive to caddy's systemd configuration.

Is there another way to to use environment variables inside my Caddyfile without modifying systemd configuration?

Best Regards !

**Comment by francislavoie** [comment]
@zieddhf seems like you're asking in totally the wrong place. This is for Caddy-Docker-Proxy, but you're using systemd. Please open a topic on the forums: https://caddy.community

**Comment by smartsys** [comment]
I hope someone can help me.
I can't get basicauth to work for phpmyadmin. 
Without the two commented out lines everything runs without problems. 

I always get the error:
ERROR: Invalid interpolation format for “labels” option in service “phpmyadmin”: “caddy.basicauth.admin:”$$2a$14$eSACJvM0K8NX9UCSzeuMFeVC5ZEwvZYeXZr9IliAXgTfDPIzNbvSC"”

I have already tested this variant and countless others.
And I already get an error if I have a space between : and ”
caddy.basicauth: "* bcrypt"
caddy.basicauth.admin: "${HASHED_PASSWORD}"



services:

  db:
    container_name: db
    image: mysql:latest
    restart: unless-stopped
    expose:
      - "3306"
    volumes:
      - ./mysql/data:/var/lib/mysql
    environment:
      MYSQL_DATABASE: ${FLASK_MYSQL_DATABASE}
      MYSQL_ROOT_PASSWORD: ${FLASK_MYSQL_ROOT_PASSWORD}
      MYSQL_USER: ${FLASK_MYSQL_USER}
      MYSQL_PASSWORD: ${FLASK_MYSQL_PASSWORD}
    networks:
      - proxy
    healthcheck:
      test: ["CMD", "mysqladmin", "ping", "-h", "localhost"]
      interval: 5s
      timeout: 5s
      retries: 3

  phpmyadmin:
    container_name: phpmyadmin
    image: phpmyadmin/phpmyadmin
    restart: unless-stopped
    expose:
      - "80"
    environment:
      PMA_HOSTS: db
      MYSQL_ROOT_PASSWORD: ${FLASK_MYSQL_ROOT_PASSWORD}
      UPLOAD_LIMIT: 128M
      PHP_UPLOAD_MAX_FILESIZE: 128M
      PHP_POST_MAX_SIZE: 128M
    depends_on:
      db:
        condition: service_healthy
    networks:
      - proxy
    labels:
      # - caddy.basicauth:"/"
      # - caddy.basicauth.admin:"$$2a$14$eSACJvM0K8NX9UCSzeuMFeVC5ZEwvZYeXZr9IliAXgTfDPIzNbvSC"
      - caddy=${PHPMYADMIN_DOMAIN}
      - caddy.reverse_proxy={{upstreams 80}}

  caddy:
    container_name: caddy
    image: lucaslorentz/caddy-docker-proxy:ci-alpine
    ports:
      - "80:80"
      - "443:443"
      - "8090:8090"
    environment:
      - CADDY_INGRESS_NETWORKS=proxy
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - ./caddy/data:/data
    networks:
      - proxy
    restart: unless-stopped

networks:
  proxy:
    name: proxy
    external: true

**Comment by u0398** [comment]
For anyone attempting workaround @francislavoie suggests, this is what I found works:

Make a Caddyfile with the snippet:
```
(basicauth) {
  basic_auth * bcrypt {
    {args[0]} {args[1]}
  }
}
```
Include the Caddyfile in the caddy-reverse-proxy container. Either with a custom build or by adding it as a compose volume. Then reference it using the `CADDY_DOCKER_CADDYFILE_PATH` environment variable:
```
volumes:
  - "<some_location>/Caddyfile:/Caddyfile"
environment:
  - CADDY_DOCKER_CADDYFILE_PATH=/Caddyfile
```
Then use a label to import the snippet in a service:
```
caddy.import: "basicauth ${BASIC_AUTH_USER} ${BASIC_AUTH_PASS}"
```
If you use an environment variable for the password make sure to wrap it in single quotes, and do not escape dollar signs ($).
```
BASIC_AUTH_USER=username
BASIC_AUTH_PASS='<hashed password>'
```

**Comment by lucaslorentz** [comment]
The root limitation (Docker Compose's variable substitution not applying to YAML keys) is outside of this project's scope, but there is a fully working workaround using a base Caddyfile with a snippet and `{args[*]}`, documented by @u0398 in [this comment](https://github.com/lucaslorentz/caddy-docker-proxy/issues/466#issuecomment-2907006017):

**Base Caddyfile** (mounted into the container and loaded via `--caddyfile-path` / `CADDY_DOCKER_CADDYFILE_PATH`):

```caddyfile
(basicauth) {
  basic_auth * bcrypt {
    {args[0]} {args[1]}
  }
}
```

**Labels** on the service that needs auth:

```yaml
labels:
  caddy: example.com
  caddy.reverse_proxy: "{{upstreams 80}}"
  caddy.import: "basicauth ${USER} ${HASHED_PASSWORD}"
```

Both `${USER}` and `${HASHED_PASSWORD}` are substituted on the YAML value side (which Compose supports), then passed as snippet arguments.

Closing since the workaround covers the original use case. If you run into a case it doesn't cover (e.g. the `{args.0}` newline parsing bug referenced earlier in this thread), please open a new issue with a minimal reproduction so it can be tracked separately.

_— Closed by Claude (AI agent) on behalf of @lucaslorentz._
